src/zcash/address/unified.cpp

```diff
@@ -75,62 +72,91 @@ ZcashdUnifiedFullViewingKey ZcashdUnifiedFullViewingKey::FromUnifiedFullViewingK
-        throw std::runtime_error("Unified addresses must include a shielded receiver.");
+        return UnifiedAddressGenerationError::ShieldedReceiverNotFound;
+    }
+
+    if (receiverTypes.count(ReceiverType::P2SH) > 0) {
+        return UnifiedAddressGenerationError::ReceiverTypeNotAvailable;
     }
 
     UnifiedAddress ua;
-    if (saplingKey.has_value() && receiverTypes.count(ReceiverType::Sapling) > 0) {
-        auto saplingAddress = saplingKey.value().Address(j);
-        if (saplingAddress.has_value()) {
-            ua.AddReceiver(saplingAddress.value());
+    if (receiverTypes.count(ReceiverType::Sapling) > 0) {
+        if (saplingKey.has_value()) {
+            auto saplingAddress = saplingKey.value().Address(j);
+            if (saplingAddress.has_value()) {
+                ua.AddReceiver(saplingAddress.value());
+            } else {
+                return UnifiedAddressGenerationError::NoAddressForDiversifier;
+            }
         } else {
-            return std::nullopt;
+            return UnifiedAddressGenerationError::ReceiverTypeNotAvailable;
         }
     }
 
-    if (transparentKey.has_value() && receiverTypes.count(ReceiverType::P2PKH) > 0) {
-        const auto& tkey = transparentKey.value();
-        auto childIndex = j.ToTransparentChildIndex();
-        if (!childIndex.has_value()) return std::nullopt;
+    if (receiverTypes.count(ReceiverType::P2PKH) > 0) {
+        if (transparentKey.has_value()) {
+            const auto& tkey = transparentKey.value();
 
-        CPubKey externalKey;
-        ChainCode externalChainCode;
-        if (!tkey.GetPubKey().Derive(externalKey, externalChainCode, 0, tkey.GetChainCode())) {
-            return std::nullopt;
-        }
+            auto childIndex = j.ToTransparentChildIndex();
+            if (!childIndex.has_value()) return UnifiedAddressGenerationError::InvalidTransparentChildIndex;
+
+            auto externalPubkey = tkey.DeriveExternal(childIndex.value());
+            if (!externalPubkey.has_value()) return UnifiedAddressGenerationError::NoAddressForDiversifier;
 
-        CPubKey childKey;
-        ChainCode childChainCode;
-        if (externalKey.Derive(childKey, childChainCode, childIndex.value(), externalChainCode)) {
-            ua.AddReceiver(childKey.GetID());
+            ua.AddReceiver(externalPubkey.value().GetID());
         } else {
-            return std::nullopt;
+            return UnifiedAddressGenerationError::ReceiverTypeNotAvailable;
         }
     }
 
-    return ua;
+    return std::make_pair(ua, j);
 }
 
-std::optional<std::pair<UnifiedAddress, diversifier_index_t>> ZcashdUnifiedFullViewingKey::FindAddress(
+UnifiedAddressGenerationResult ZcashdUnifiedFullViewingKey::FindAddress(
         const diversifier_index_t& j,
         const std::set<ReceiverType>& receiverTypes) const {
     diversifier_index_t j0(j);
     bool hasTransparent = HasTransparent(receiverTypes);
     auto addr = Address(j0, receiverTypes);
```

Comment:
The `hasTransparent` check is actually superfluous, since the `Address` call does that internally anyway; making it so that this only loops if it is not `NoAddressForDiversifier` solves the whole problem, I think, and avoids a redundant check of j0.